src/context/distributor.hpp

```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include "cocaine/memory.hpp"
+#include "cocaine/dynamic.hpp"
+#include "cocaine/errors.hpp"
+
+namespace cocaine {
+
+// distributor of elements of Pool
+// Pool should provide value_type typedef and 'operator[](size_t)' call
+// value_type should provide operator* and oprator->,
+// resulting type should provide 'double utilization()' call with returned values in [0.0 1.0] range
+template <class Pool>
+struct distributor {
+    using result_type = typename Pool::value_type;
+
+    virtual
+    ~distributor() {}
+
+    virtual
+    auto
+    next(Pool& pool) -> result_type& = 0;
+};
+
+
+// Distributes elements of the Pool according only to 'utilization()' of the elements.
+// Simply returns the lowest value.
+template <class Pool>
+struct load_distributor: public distributor<Pool> {
+    using result_type = typename Pool::value_type;
+
+    load_distributor() {}
+
+    auto
+    next(Pool& pool) -> result_type& override {
+        auto comp = [](const result_type& lhs, const result_type& rhs) {
+            return lhs->utilization() < rhs->utilization();
+        };
+        return *std::min_element(pool.begin(), pool.end(), comp);
+    }
+};
+
+// Distributes elements of the Pool according to primitive round robin algorithm
+// Always returns element with next index (in a circular manner)
+template <class Pool>
+struct rr_distributor: public distributor<Pool> {
+    using result_type = typename Pool::value_type;
+
+    rr_distributor() : counter(0) {}
+
+    auto
+    next(Pool& pool) -> result_type& override {
+        return pool[counter++ % pool.size()];
+    }
+
+private:
+    std::atomic_uint counter;
+};
+
+// Split elements to several buckets according to utilization
+// Chooses bucket with lowest value and returns random element from it
+template <class Pool>
+struct bucket_random_distributor: public distributor<Pool> {
+    using result_type = typename Pool::value_type;
+    using bucket_t = std::vector<std::reference_wrapper<result_type>>;
+
+    static constexpr const double max_value = 1.0;
+
+    bucket_random_distributor(const dynamic_t& args) :
+        bucket_size(args.as_object().at("bucket_size", 0.02).as_double())
+    {}
+
+    auto
+    next(Pool& pool) -> result_type& override {
+        auto units = get_bucket_with_minimal_utilization(pool);
+        return units[rand() % units.size()];
```

Comment:
Is it ok to obtain empty `units` sequence here?

Reply:
Is it `std::rand` here? What guaranties are on maximum random value?

Reply:
units should not be empty as far as pool is not empty. 
yep, it's std::rand, I think it's good enough for us returning values in range of [0, max int32]